stan/math/prim/prob/hmm_marginal_lpdf.hpp

```diff
@@ -0,0 +1,187 @@
+  eig_vector_partial alpha_log_norms(n_transitions + 1);
+  eig_matrix_partial omegas;
+  auto Gamma_val = value_of(Gamma).eval();
+
+  // compute the density using the forward algorithm.
+  {
+    const auto log_omegas_val = value_of(log_omegas).eval();
+    const auto rho_val = value_of(rho).eval();
+    omegas = log_omegas_val.array().exp();
+    const int n_states = log_omegas_val.rows();
+    const int n_transitions = log_omegas_val.cols() - 1;
+
+    alphas.col(0) = omegas.col(0).cwiseProduct(rho_val);
+
+    const auto norm = alphas.col(0).maxCoeff();
+    alphas.col(0) /= norm;
+    alpha_log_norms(0) = log(norm);
+
+    for (int n = 0; n < n_transitions; ++n) {
+      alphas.col(n + 1) = omegas.col(n + 1).cwiseProduct(Gamma_val.transpose()
+                                                         * alphas.col(n));
+
+      const auto col_norm = alphas.col(n + 1).maxCoeff();
+      alphas.col(n + 1) /= col_norm;
+      alpha_log_norms(n + 1) = log(col_norm) + alpha_log_norms(n);
+    }
+  }
+  const auto log_marginal_density
+      = log(alphas.col(n_transitions).sum()) + alpha_log_norms(n_transitions);
+
+  // Variables required for all three Jacobian-adjoint products.
+  const auto norm_norm = alpha_log_norms(n_transitions);
+  const auto unnormed_marginal = alphas.col(n_transitions).sum();
+
+  std::vector<eig_vector_partial> kappa(n_transitions);
+  eig_vector_partial kappa_log_norms(n_transitions);
+  std::vector<T_partial_type> grad_corr(n_transitions, 0);
+
+  if (n_transitions > 0) {
+    kappa[n_transitions - 1] = Eigen::VectorXd::Ones(n_states);
+    kappa_log_norms(n_transitions - 1) = 0;
+    grad_corr[n_transitions - 1]
+        = exp(alpha_log_norms(n_transitions - 1) - norm_norm);
+  }
+
+  for (int n = n_transitions - 2; n >= 0; --n) {
+    kappa[n] = Gamma_val * (omegas.col(n + 2).cwiseProduct(kappa[n + 1]));
+
+    const auto norm = kappa[n].maxCoeff();
+    kappa[n] /= norm;
+    kappa_log_norms[n] = log(norm) + kappa_log_norms[n + 1];
+    grad_corr[n] = exp(alpha_log_norms[n] + kappa_log_norms[n] - norm_norm);
+  }
+
+  if (!is_constant_all<T_Gamma>::value) {
+    eig_matrix_partial Gamma_jacad = Eigen::MatrixXd::Zero(n_states, n_states);
+
+    for (int n = n_transitions - 1; n >= 0; --n) {
+      Gamma_jacad += (grad_corr[n] * kappa[n].cwiseProduct(omegas.col(n + 1))
+                      * alphas.col(n).transpose())
+                         .transpose();
+    }
+
+    Gamma_jacad /= unnormed_marginal;
+    ops_partials.edge2_.partials_ = Gamma_jacad;
+  }
+
+  if (!is_constant_all<T_omega>::value || !is_constant_all<T_rho>::value) {
+    eig_matrix_partial log_omega_jacad
+        = Eigen::MatrixXd::Zero(n_states, n_transitions + 1);
+
+    if (!is_constant_all<T_omega>::value) {
+      for (int n = n_transitions - 1; n >= 0; --n)
+        log_omega_jacad.col(n + 1)
+            = grad_corr[n]
+              * kappa[n].cwiseProduct(Gamma_val.transpose() * alphas.col(n));
```

Comment:
That's a good point. Changing this up causes issues with fwd mode differentiation. It's still unclear to me why fwd diff adds these complications. Here's my attempt:
```
    if (!is_constant_all<T_omega>::value) {
      auto Gamma_alpha = Gamma_val.transpose() * alphas;
      for (int n = n_transitions - 1; n >= 0; --n)
        log_omega_jacad.col(n + 1)
            = grad_corr[n] // *  elt_multiply(kappa[n], Gamma_alpha.col(n));
              * kappa[n].cwiseProduct(Gamma_alpha.col(n)).eval();
}
```

The error message is
```
lib/eigen_3.3.3/Eigen/src/Core/products/GeneralBlockPanelKernel.h:1555:23: error: 
      no matching member function for call to 'madd'
              straits.madd(A0,B_0,C0,B_0);
              ~~~~~~~~^~~~
lib/eigen_3.3.3/Eigen/src/Core/products/GeneralMatrixMatrix.h:194:11: note: in
      instantiation of member function 'Eigen::internal::gebp_kernel<double,
      stan::math::fvar<double>, long,
      Eigen::internal::blas_data_mapper<stan::math::fvar<double>, long, 0, 0>,
      2, 4, false, false>::operator()' requested here
          gebp(res.getSubMapper(i2, j2), blockA, blockB, actual_mc, actu...
          ^
lib/eigen_3.3.3/Eigen/src/Core/products/GeneralMatrixMatrix.h:226:11: note: in
      instantiation of member function
      'Eigen::internal::general_matrix_matrix_product<long, double, 1, false,
      stan::math::fvar<double>, 0, false, 0>::run' requested here
    Gemm::run(rows, cols, m_lhs.cols(),
          ^
lib/eigen_3.3.3/Eigen/src/Core/products/Parallelizer.h:97:3: note: in
      instantiation of member function
      'Eigen::internal::gemm_functor<stan::math::fvar<double>, long,
      Eigen::internal::general_matrix_matrix_product<long, double, 1, false,
      stan::math::fvar<double>, 0, false, 0>, Eigen::Transpose<const
      Eigen::Matrix<double, -1, -1, 0, -1, -1> >,
      Eigen::Matrix<stan::math::fvar<double>, -1, -1, 0, -1, -1>,
      Eigen::Matrix<stan::math::fvar<double>, -1, -1, 0, -1, -1>,
      Eigen::internal::gemm_blocking_space<0, double, stan::math::fvar<double>,
      -1, -1, -1, 1, false> >::operator()' requested here
  func(0,rows, 0,cols);
  ^
lib/eigen_3.3.3/Eigen/src/Core/products/GeneralMatrixMatrix.h:483:15: note: in
      instantiation of function template specialization
      'Eigen::internal::parallelize_gemm<true,
      Eigen::internal::gemm_functor<stan::math::fvar<double>, long,
      Eigen::internal::general_matrix_matrix_product<long, double, 1, false,
      stan::math::fvar<double>, 0, false, 0>, Eigen::Transpose<const
      Eigen::Matrix<double, -1, -1, 0, -1, -1> >,
      Eigen::Matrix<stan::math::fvar<double>, -1, -1, 0, -1, -1>,
      Eigen::Matrix<stan::math::fvar<double>, -1, -1, 0, -1, -1>,
      Eigen::internal::gemm_blocking_space<0, double, stan::math::fvar<double>,
      -1, -1, -1, 1, false> >, long>' requested here
    internal::parallelize_gemm<(Dest::MaxRowsAtCompileTime>32 || Dest::M...
              ^
lib/eigen_3.3.3/Eigen/src/Core/products/GeneralMatrixMatrix.h:435:7: note: in
      instantiation of function template specialization
      'Eigen::internal::generic_product_impl<Eigen::Transpose<Eigen::Matrix<double,
      -1, -1, 0, -1, -1> >, Eigen::Matrix<stan::math::fvar<double>, -1, -1, 0,
      -1, -1>, Eigen::DenseShape, Eigen::DenseShape,
      8>::scaleAndAddTo<Eigen::Matrix<stan::math::fvar<double>, -1, -1, 0, -1,
      -1> >' requested here
      scaleAndAddTo(dst, lhs, rhs, Scalar(1));
      ^
lib/eigen_3.3.3/Eigen/src/Core/ProductEvaluators.h:124:69: note: in
      instantiation of function template specialization
      'Eigen::internal::generic_product_impl<Eigen::Transpose<Eigen::Matrix<double,
      -1, -1, 0, -1, -1> >, Eigen::Matrix<stan::math::fvar<double>, -1, -1, 0,
      -1, -1>, Eigen::DenseShape, Eigen::DenseShape,
      8>::evalTo<Eigen::Matrix<stan::math::fvar<double>, -1, -1, 0, -1, -1> >'
      requested here
    generic_product_impl<Lhs, Rhs, LhsShape, RhsShape, ProductTag>::evalTo(m...
                                                                    ^
lib/eigen_3.3.3/Eigen/src/Core/ProductEvaluators.h:35:62: note: in instantiation
      of member function
      'Eigen::internal::product_evaluator<Eigen::Product<Eigen::Transpose<Eigen::Matrix<double,
      -1, -1, 0, -1, -1> >, Eigen::Matrix<stan::math::fvar<double>, -1, -1, 0,
      -1, -1>, 0>, 8, Eigen::DenseShape, Eigen::DenseShape, double,
      stan::math::fvar<double> >::product_evaluator' requested here
  EIGEN_DEVICE_FUNC explicit evaluator(const XprType& xpr) : Base(xpr) {}
                                                             ^
lib/eigen_3.3.3/Eigen/src/Core/CoreEvaluators.h:1019:7: note: in instantiation
      of member function
      'Eigen::internal::evaluator<Eigen::Product<Eigen::Transpose<Eigen::Matrix<double,
      -1, -1, 0, -1, -1> >, Eigen::Matrix<stan::math::fvar<double>, -1, -1, 0,
      -1, -1>, 0> >::evaluator' requested here
    : m_argImpl(block.nestedExpression()), 
      ^
lib/eigen_3.3.3/Eigen/src/Core/CoreEvaluators.h:1008:7: note: in instantiation
      of member function
      'Eigen::internal::unary_evaluator<Eigen::Block<Eigen::Product<Eigen::Transpose<Eigen::Matrix<double,
      -1, -1, 0, -1, -1> >, Eigen::Matrix<stan::math::fvar<double>, -1, -1, 0,
      -1, -1>, 0>, -1, 1, true>, Eigen::internal::IndexBased,
      stan::math::fvar<double> >::unary_evaluator' requested here
    : unary_evaluator<XprType>(block) 
      ^
lib/eigen_3.3.3/Eigen/src/Core/CoreEvaluators.h:994:64: note: in instantiation
      of member function
      'Eigen::internal::block_evaluator<Eigen::Product<Eigen::Transpose<Eigen::Matrix<double,
      -1, -1, 0, -1, -1> >, Eigen::Matrix<stan::math::fvar<double>, -1, -1, 0,
      -1, -1>, 0>, -1, 1, true, false>::block_evaluator' requested here
  EIGEN_DEVICE_FUNC explicit evaluator(const XprType& block) : block_eva...
                                                               ^
lib/eigen_3.3.3/Eigen/src/Core/CoreEvaluators.h:103:38: note: in instantiation
      of member function
      'Eigen::internal::evaluator<Eigen::Block<Eigen::Product<Eigen::Transpose<Eigen::Matrix<double,
      -1, -1, 0, -1, -1> >, Eigen::Matrix<stan::math::fvar<double>, -1, -1, 0,
      -1, -1>, 0>, -1, 1, true> >::evaluator' requested here
  explicit evaluator(const T& xpr) : evaluator<T>(xpr) {}
                                     ^
lib/eigen_3.3.3/Eigen/src/Core/CoreEvaluators.h:702:7: note: in instantiation of
      member function 'Eigen::internal::evaluator<const
      Eigen::Block<Eigen::Product<Eigen::Transpose<Eigen::Matrix<double, -1, -1,
      0, -1, -1> >, Eigen::Matrix<stan::math::fvar<double>, -1, -1, 0, -1, -1>,
      0>, -1, 1, true> >::evaluator' requested here
      m_rhsImpl(xpr.rhs())  
      ^
lib/eigen_3.3.3/Eigen/src/Core/CoreEvaluators.h:671:62: note: in instantiation
      of member function
      'Eigen::internal::binary_evaluator<Eigen::CwiseBinaryOp<Eigen::internal::scalar_product_op<stan::math::fvar<double>,
      stan::math::fvar<double> >, const Eigen::Matrix<stan::math::fvar<double>,
      -1, 1, 0, -1, 1>, const
      Eigen::Block<Eigen::Product<Eigen::Transpose<Eigen::Matrix<double, -1, -1,
      0, -1, -1> >, Eigen::Matrix<stan::math::fvar<double>, -1, -1, 0, -1, -1>,
      0>, -1, 1, true> >, Eigen::internal::IndexBased,
      Eigen::internal::IndexBased, stan::math::fvar<double>,
      stan::math::fvar<double> >::binary_evaluator' requested here
  EIGEN_DEVICE_FUNC explicit evaluator(const XprType& xpr) : Base(xpr) {}
                                                             ^
lib/eigen_3.3.3/Eigen/src/Core/AssignEvaluator.h:730:20: note: in instantiation
      of member function
      'Eigen::internal::evaluator<Eigen::CwiseBinaryOp<Eigen::internal::scalar_product_op<stan::math::fvar<double>,
      stan::math::fvar<double> >, const Eigen::Matrix<stan::math::fvar<double>,
      -1, 1, 0, -1, 1>, const
      Eigen::Block<Eigen::Product<Eigen::Transpose<Eigen::Matrix<double, -1, -1,
      0, -1, -1> >, Eigen::Matrix<stan::math::fvar<double>, -1, -1, 0, -1, -1>,
      0>, -1, 1, true> > >::evaluator' requested here
  SrcEvaluatorType srcEvaluator(src);
                   ^
lib/eigen_3.3.3/Eigen/src/Core/AssignEvaluator.h:879:5: note: in instantiation
      of function template specialization
      'Eigen::internal::call_dense_assignment_loop<Eigen::Matrix<stan::math::fvar<double>,
      -1, 1, 0, -1, 1>,
      Eigen::CwiseBinaryOp<Eigen::internal::scalar_product_op<stan::math::fvar<double>,
      stan::math::fvar<double> >, const Eigen::Matrix<stan::math::fvar<double>,
      -1, 1, 0, -1, 1>, const
      Eigen::Block<Eigen::Product<Eigen::Transpose<Eigen::Matrix<double, -1, -1,
      0, -1, -1> >, Eigen::Matrix<stan::math::fvar<double>, -1, -1, 0, -1, -1>,
      0>, -1, 1, true> >, Eigen::internal::assign_op<stan::math::fvar<double>,
      stan::math::fvar<double> > >' requested here
    call_dense_assignment_loop(dst, src, func);
    ^
lib/eigen_3.3.3/Eigen/src/Core/AssignEvaluator.h:836:46: note: in instantiation
      of member function
      'Eigen::internal::Assignment<Eigen::Matrix<stan::math::fvar<double>, -1,
      1, 0, -1, 1>,
      Eigen::CwiseBinaryOp<Eigen::internal::scalar_product_op<stan::math::fvar<double>,
      stan::math::fvar<double> >, const Eigen::Matrix<stan::math::fvar<double>,
      -1, 1, 0, -1, 1>, const
      Eigen::Block<Eigen::Product<Eigen::Transpose<Eigen::Matrix<double, -1, -1,
      0, -1, -1> >, Eigen::Matrix<stan::math::fvar<double>, -1, -1, 0, -1, -1>,
      0>, -1, 1, true> >, Eigen::internal::assign_op<stan::math::fvar<double>,
      stan::math::fvar<double> >, Eigen::internal::Dense2Dense, void>::run'
      requested here
  Assignment<ActualDstTypeCleaned,Src,Func>::run(actualDst, src, func);
                                             ^
lib/eigen_3.3.3/Eigen/src/Core/PlainObjectBase.h:728:17: note: in instantiation
      of function template specialization
      'Eigen::internal::call_assignment_no_alias<Eigen::Matrix<stan::math::fvar<double>,
      -1, 1, 0, -1, 1>,
      Eigen::CwiseBinaryOp<Eigen::internal::scalar_product_op<stan::math::fvar<double>,
      stan::math::fvar<double> >, const Eigen::Matrix<stan::math::fvar<double>,
      -1, 1, 0, -1, 1>, const
      Eigen::Block<Eigen::Product<Eigen::Transpose<Eigen::Matrix<double, -1, -1,
      0, -1, -1> >, Eigen::Matrix<stan::math::fvar<double>, -1, -1, 0, -1, -1>,
      0>, -1, 1, true> >, Eigen::internal::assign_op<stan::math::fvar<double>,
      stan::math::fvar<double> > >' requested here
      internal::call_assignment_no_alias(this->derived(), other.derived(...
                ^
lib/eigen_3.3.3/Eigen/src/Core/PlainObjectBase.h:812:13: note: in instantiation
      of function template specialization
      'Eigen::PlainObjectBase<Eigen::Matrix<stan::math::fvar<double>, -1, 1, 0,
      -1, 1>
      >::_set_noalias<Eigen::CwiseBinaryOp<Eigen::internal::scalar_product_op<stan::math::fvar<double>,
      stan::math::fvar<double> >, const Eigen::Matrix<stan::math::fvar<double>,
      -1, 1, 0, -1, 1>, const
      Eigen::Block<Eigen::Product<Eigen::Transpose<Eigen::Matrix<double, -1, -1,
      0, -1, -1> >, Eigen::Matrix<stan::math::fvar<double>, -1, -1, 0, -1, -1>,
      0>, -1, 1, true> > >' requested here
      this->_set_noalias(other);
            ^
lib/eigen_3.3.3/Eigen/src/Core/Matrix.h:296:22: note: in instantiation of
      function template specialization
      'Eigen::PlainObjectBase<Eigen::Matrix<stan::math::fvar<double>, -1, 1, 0,
      -1, 1>
      >::_init1<Eigen::CwiseBinaryOp<Eigen::internal::scalar_product_op<stan::math::fvar<double>,
      stan::math::fvar<double> >, const Eigen::Matrix<stan::math::fvar<double>,
      -1, 1, 0, -1, 1>, const
      Eigen::Block<Eigen::Product<Eigen::Transpose<Eigen::Matrix<double, -1, -1,
      0, -1, -1> >, Eigen::Matrix<stan::math::fvar<double>, -1, -1, 0, -1, -1>,
      0>, -1, 1, true> >,
      Eigen::CwiseBinaryOp<Eigen::internal::scalar_product_op<stan::math::fvar<double>,
      stan::math::fvar<double> >, const Eigen::Matrix<stan::math::fvar<double>,
      -1, 1, 0, -1, 1>, const
      Eigen::Block<Eigen::Product<Eigen::Transpose<Eigen::Matrix<double, -1, -1,
      0, -1, -1> >, Eigen::Matrix<stan::math::fvar<double>, -1, -1, 0, -1, -1>,
      0>, -1, 1, true> > >' requested here
      Base::template _init1<T>(x);
                     ^
lib/eigen_3.3.3/Eigen/src/Core/DenseBase.h:406:14: note: in instantiation of
      function template specialization 'Eigen::Matrix<stan::math::fvar<double>,
      -1, 1, 0, -1,
      1>::Matrix<Eigen::CwiseBinaryOp<Eigen::internal::scalar_product_op<stan::math::fvar<double>,
      stan::math::fvar<double> >, const Eigen::Matrix<stan::math::fvar<double>,
      -1, 1, 0, -1, 1>, const
      Eigen::Block<Eigen::Product<Eigen::Transpose<Eigen::Matrix<double, -1, -1,
      0, -1, -1> >, Eigen::Matrix<stan::math::fvar<double>, -1, -1, 0, -1, -1>,
      0>, -1, 1, true> > >' requested here
      return typename internal::eval<Derived>::type(derived());
             ^
./stan/math/prim/prob/hmm_marginal_lpdf.hpp:156:59: note: in instantiation of
      member function
      'Eigen::DenseBase<Eigen::CwiseBinaryOp<Eigen::internal::scalar_product_op<stan::math::fvar<double>,
      stan::math::fvar<double> >, const Eigen::Matrix<stan::math::fvar<double>,
      -1, 1, 0, -1, 1>, const
      Eigen::Block<Eigen::Product<Eigen::Transpose<Eigen::Matrix<double, -1, -1,
      0, -1, -1> >, Eigen::Matrix<stan::math::fvar<double>, -1, -1, 0, -1, -1>,
      0>, -1, 1, true> > >::eval' requested here
              * kappa[n].cwiseProduct(Gamma_alpha.col(n)).eval();
                                                          ^
test/unit/math/prim/prob/hmm_marginal_test.cpp:122:12: note: in instantiation of
      function template specialization
      'hmm_marginal_test_wrapper<stan::math::fvar<stan::math::fvar<double> >,
      double, double>' requested here
    return hmm_marginal_test_wrapper(log_omegas, Gamma_unconstrained,
           ^
./test/unit/math/test_ad.hpp:372:3: note: in instantiation of function template
      specialization
      'stan::test::internal::expect_throw<stan::math::fvar<stan::math::fvar<double>
      >, (lambda at ./test/unit/math/test_ad.hpp:453:35)>' requested here
  expect_throw<fvar<fvar<double>>>(f, x, "fvar<fvar<double>>");
  ^
./test/unit/math/test_ad.hpp:462:15: note: in instantiation of function template
      specialization 'stan::test::internal::expect_all_throw<(lambda at
      ./test/unit/math/test_ad.hpp:453:35)>' requested here
    internal::expect_all_throw(h(0), x);
              ^
./test/unit/math/test_ad.hpp:662:13: note: in instantiation of function template
      specialization 'stan::test::internal::expect_ad_helper<(lambda at
      test/unit/math/prim/prob/hmm_marginal_test.cpp:120:22), (lambda at
      ./test/unit/math/test_ad.hpp:657:13), Eigen::Matrix<double, -1, -1, 0, -1,
      -1>, Eigen::Matrix<double, -1, -1, 0, -1, -1>, std::__1::vector<double,
      std::__1::allocator<double> > >' requested here
  internal::expect_ad_helper(tols, f, g1, serialize_args(x1), x1, x2, x3);
            ^
./test/unit/math/test_ad.hpp:1193:13: note: in instantiation of function
      template specialization 'stan::test::internal::expect_ad_vvv<(lambda at
      test/unit/math/prim/prob/hmm_marginal_test.cpp:120:22),
      Eigen::Matrix<double, -1, -1, 0, -1, -1>, Eigen::Matrix<double, -1, -1, 0,
      -1, -1>, std::__1::vector<double, std::__1::allocator<double> > >'
      requested here
  internal::expect_ad_vvv(tols, f, x1, x2, x3);
            ^
test/unit/math/prim/prob/hmm_marginal_test.cpp:126:15: note: in instantiation of
      function template specialization 'stan::test::expect_ad<(lambda at
      test/unit/math/prim/prob/hmm_marginal_test.cpp:120:22),
      Eigen::Matrix<double, -1, -1, 0, -1, -1>, Eigen::Matrix<double, -1, -1, 0,
      -1, -1>, std::__1::vector<double, std::__1::allocator<double> > >'
      requested here
  stan::test::expect_ad(tols, hmm_functor, log_omegas, Gamma_unconstrained,
              ^
lib/eigen_3.3.3/Eigen/src/Core/products/GeneralBlockPanelKernel.h:435:28: note: 
      candidate template ignored: deduced conflicting types for parameter
      'AccPacketType' ('stan::math::fvar<double>' vs. 'double')
  EIGEN_STRONG_INLINE void madd(const LhsPacketType& a, const RhsPacketT...                       
```

It gets repeated with slight variations because the unit tests call the function multipe times.

Reply:
The problem is missing matrix arithmetic operators for fvar.  We also have some missing for var.  Anyway, no big deal---we can try to make local expressions more efficient later.

Reply:
Ok, I'll leave the code as is for now.